File: apps/macos/Sources/WispProjectBrowser/NativeProvenance.swift
import Foundation

/// Projection of the displayed transcript, matching WebView's tool-only provenance list.
/// The index is scoped to one transcript page; hosts reset row UI state on page changes.
public struct NativeProvenanceRow: Equatable, Identifiable, Sendable {
    public let index: Int
    public let name: String
    public let input: String
    public let output: String
    public let ok: Bool?
    public var id: String { "\(index):\(name)" }
    public var initiallyExpanded: Bool { ok != true }
    public static func collect(_ items: [ConversationItem]) -> [Self] {
        items.enumerated().compactMap { index, item in
            guard item.role == "tool" else { return nil }
            return Self(index: index, name: item.tool_name ?? "", input: item.input ?? "", output: item.text, ok: item.ok)
        }
    }
    public func matches(_ query: String) -> Bool {
        query.isEmpty || [name, input, output].contains { $0.localizedCaseInsensitiveContains(query) }
    }
}

File: apps/windows/Wisp.ProjectBrowser.Contracts/NativeProvenance.cs
namespace Wisp.ProjectBrowser.Contracts;

/// A view over the displayed transcript, not a separate backend store.
/// Reset disclosure state when switching transcript pages; preserve it during live updates.
public sealed record NativeProvenanceRow(int Index, string Name, string Input, string Output, bool? Ok)
{
    public string Id => $"{Index}:{Name}";
    public bool InitiallyExpanded => Ok != true;
    public bool Matches(string query) => query.Length == 0 || new[] { Name, Input, Output }.Any(value => value.Contains(query, StringComparison.CurrentCultureIgnoreCase));
    public static NativeProvenanceRow[] Collect(IEnumerable<ConversationItem> items) => items
        .Select((item, index) => (item, index)).Where(pair => pair.item.Role == "tool")
        .Select(pair => new NativeProvenanceRow(pair.index, pair.item.ToolName ?? "", pair.item.Input ?? "", pair.item.Text, pair.item.Ok)).ToArray();
}
